redis/connection.py

```diff
@@ -538,9 +540,10 @@ def __init__(
         self.host = host
         self.port = int(port)
         self.db = db
-        self.username = username
         self.client_name = client_name
-        self.password = password
+        self.credentials_provider = credentials_provider
+        if not self.credentials_provider and (username or password):
```

Comment:
Can we order change? Positive ifs are easier to read than negatives - to be kind to our community.